core/services/ocr2/plugins/ocr2keeper/evm21/registry.go

```diff
@@ -277,27 +280,93 @@ func (r *EvmRegistry) refreshActiveUpkeeps() error {
+		end := i + logTriggerRefreshBatchSize
+		if end > len(ids) {
+			end = len(ids)
+		}
+		batch := ids[i:end]
+
+		if batchErr := r.refreshLogTriggerUpkeepsBatch(batch); batchErr != nil {
+			multierr.AppendInto(&err, batchErr)
+		}
+
+		time.Sleep(500 * time.Millisecond)
+	}
+
+	return err
+}
+
+func (r *EvmRegistry) refreshLogTriggerUpkeepsBatch(ids []*big.Int) error {
+	var logTriggerIDs []*big.Int
+	var logTriggerHashes []common.Hash
 	for _, id := range ids {
 		uid := &ocr2keepers.UpkeepIdentifier{}
 		if ok := uid.FromBigInt(id); !ok {
 			r.lggr.Warnf("failed to parse upkeep id %s", id.String())
+			continue
 		}
 		switch core.GetUpkeepType(*uid) {
 		case ocr2keepers.LogTrigger:
 			logTriggerIDs = append(logTriggerIDs, id)
-		default:
+			logTriggerHashes = append(logTriggerHashes, common.BigToHash(id))
 		}
 	}
 	newUpkeeps, err := r.logEventProvider.RefreshActiveUpkeeps(logTriggerIDs...)
 	if err != nil {
 		return fmt.Errorf("failed to refresh active upkeep ids in log event provider: %w", err)
 	}
+
+	unpausedLogs, err := r.poller.IndexedLogs(iregistry21.IKeeperRegistryMasterUpkeepUnpaused{}.Topic(), r.addr, 1, logTriggerHashes, indexedLogsConfirmations, pg.WithParentCtx(r.ctx))
+	if err != nil {
+		return err
+	}
+	configSetLogs, err := r.poller.IndexedLogs(iregistry21.IKeeperRegistryMasterUpkeepTriggerConfigSet{}.Topic(), r.addr, 1, logTriggerHashes, indexedLogsConfirmations, pg.WithParentCtx(r.ctx))
+	if err != nil {
+		return err
+	}
+
+	logs := append(unpausedLogs, configSetLogs...)
+
+	configSetBlockNumbers := map[string]uint64{}
+	unpausedBlockNumbers := map[string]uint64{}
+	perUpkeepConfig := map[string][]byte{}
+
+	for _, log := range logs {
+		rawLog := log.ToGethLog()
+		abilog, err := r.registry.ParseLog(rawLog)
+		if err != nil {
+			return err
+		}
+		switch l := abilog.(type) {
+		case *iregistry21.IKeeperRegistryMasterUpkeepTriggerConfigSet:
+			configSetBlockNumbers[l.Id.String()] = rawLog.BlockNumber
+			perUpkeepConfig[l.Id.String()] = l.TriggerConfig
+		case *iregistry21.IKeeperRegistryMasterUpkeepUnpaused:
+			unpausedBlockNumbers[l.Id.String()] = rawLog.BlockNumber
+		}
+	}
+
 	var merr error
 	for _, id := range newUpkeeps {
-		// TODO: find the ConfigSet/UpkeepUnpaused events for this upkeep and pass cfg and block number
-		// block number should be taken from UpkeepUnpaused if it's block is higher than ConfigSet
-		if err := r.updateTriggerConfig(id, nil, 0); err != nil {
+		logBlock, ok := configSetBlockNumbers[id.String()]
+		if !ok {
+			r.lggr.Warnf("unable to find config set block number for %s", id.String())
+			continue
+		}
+
+		config, ok := perUpkeepConfig[id.String()]
+		if !ok {
+			r.lggr.Warnf("unable to find per upkeep config for %s", id.String())
+			continue
+		}
+
+		// In case an upkeep was paused then unpaused after a config set event, start the config from the unpaused block number
+		if unpausedBlockNumbers[id.String()] > logBlock {
```

Comment:
curious, will `unpausedBlockNumbers[id.String()]` return 0 if id.String() was not found?

Reply:
Yes that's correct